### ISSUE TITLE
Make an alternative interface without hidden shared state

### ISSUE THREAD

**Issue by wizardsd** [nanomsg/nng#737]
## NNG & Platform details.

NNG v1.0.1, Windows

## Expected Behavior

There is no shared state in backend

## Actual Behavior

There are static variables and I have to manage their state manually (create and destroy).

I just dislike singletons, they are unusable in enterprise production code (from my experience).  I'd prefer a library without hidden shared state. Could you create an alternative API to use nng without singletons?

For example: what if I have a C++ class that uses NNG, each instance should shutdown NNG (C++ programmers love RAII), but with current implementation shutting down NNG in one instance makes it unusable in second instance of the class. What if multiple C++ libraries depends on NNG and do not interact with each other in the same process but have to be shutdown properly?

https://github.com/nanomsg/nng/blob/master/src/platform/windows/win_io.c#L22
https://stackoverflow.com/a/138012

**Comment by gdamore** [comment]
In general you should not ever need to shutdown the entire library.  The
only reason to do so is to silence warnings from leak analysis tools
because of global tables.

Instead just open and close sockets and stop worrying about the global
tables.

 - Garrett

**Comment by gdamore** [comment]
To put a little more meaningful content here.  The nanomsg API and NNG are
designed around the idea that sockets are just indices into a table much
like file descriptors.  This is natural for C.

In order to make this work we need to hold those tables somewhere.  We
could instead pass a global state object with those tables everywhere but
that would be unnatural.

In C++ the natural thing to do would be to model the sockets as objects and
not even have the tables.  In Go I did just this in my mangos library (no
global state).

NNG was built to satisfy C paradigms and there is a bit of an impedance
mismatch for C++ developers.

I can imagine that it would be desirable someday to make a pure C++ version
of NNG that didn't have any of that state. As I am not a C++ user anymore
(or only very infrequently) this hasn't been a priority.

If some company wanted to fund a pure C++ rewrite I would be happy to help.
But absent that, C++ users have to live with the lowest common denominator
of C99 when facing NNG.  Sorry.

 - Garrett

**Comment by wizardsd** [comment]
Yes, you are right, I'm from C++ world. You know, that when you keep singletons alive they will be shut down in undefined order. In C++ we actually like '}' for defining what and when will be destroyed. And I like "What you don’t use, you don’t pay for" idea of Bjarne Stroustrup, if i don't need a socket any more, why should I pay for it's backend? Maybe you could somehow incorporate pointer to shared state to socket handle and make something like reference counting + Phoenix singleton for shared state?

Absence of that somehow limits usage of NNG because I'd prefer independent tests of my frontend, but with shared state in NNG my tests will depend somehow on each other in unpredictable way. Sometimes I prefer to run tests in parallel (one test per thread for example) on CI server.

The second, from my point of view, that user of NNG shouldn't know about that shared state in backend. I've seen many times junior programmers that are calling `WSACleanup`, `fcloseall` in their applications while the application is still running and wonder of unexpected effects... so for me it would be better not to have such public API at all (maybe in some private headers just for testing purposes only). For me it looks like big red button with a caption "Don't press it" (and you know what will happen one day).

**Comment by gdamore** [comment]
I thing for nng_fini I have pretty clearly indicated that it's use is
discouraged and if it is used at all it should be done from atexit(). It
exists solely to clean up the few global tables that *must* exist, and that
only for people using tools like valgrind. This function prevents a false
positive leak report.

The reason the table exists is to permit lookup of an object by id, and to
prevent I'd reuse. This also makes it much easier to avoid use after free
bugs in end user code.

For example if we used a pointer instead of the identifier then it would be
extremely difficult or impossible to make close() a threadsafe operation.
You would need to introduce an extra destructor function. C++ automatic
scope can make that less terrible but that's not available for C.  And even
for C++ care must be taken not to destroy a heap object when it may be in
use from another thread.

Integer identifiers instead make this whole mess a non issue. But the cost
is that global table behind the scenes.

In a garbage collected language this is not an issue.  But neither C nor
C++ have gc.

Upshot is that if the presence of global state offends you then maybe this
isn't the library for you. I won't be changing this aspect of the code.

I believe zeromq has a global context object that is explicitly created,
passed, and destroyed by api users.  You might find that more to your
liking. In our experience we found that undesirable and a barrier to
adoption. But you might prefer that.

  - Garrett
